titiler/stacapi/asset_reader.py

```diff
@@ -0,0 +1,222 @@
+    ExpressionMixingWarning,
+    InvalidAssetName,
+    MissingAssets,
+    TileOutsideBounds,
+)
+from rio_tiler.io import Reader
+from rio_tiler.io.base import BaseReader, MultiBaseReader
+from rio_tiler.models import ImageData
+from rio_tiler.tasks import multi_arrays
+from rio_tiler.types import Indexes
+
+from titiler.stacapi.models import AssetInfo
+from titiler.stacapi.settings import CacheSettings, RetrySettings, STACSettings
+from titiler.stacapi.xarray import XarrayReader
+
+cache_config = CacheSettings()
+retry_config = RetrySettings()
+stac_config = STACSettings()
+
+valid_types = {
+    "image/tiff; application=geotiff",
+    "image/tiff; application=geotiff; profile=cloud-optimized",
+    "image/tiff; profile=cloud-optimized; application=geotiff",
+    "image/vnd.stac.geotiff; cloud-optimized=true",
+    "image/tiff",
+    "image/x.geotiff",
+    "image/jp2",
+    "application/x-hdf5",
+    "application/x-hdf",
+    "application/vnd+zarr",
+    "application/x-netcdf",
+}
+
+
+@attr.s
+class AssetReader(MultiBaseReader):
+    """
+    Asset reader for STAC items.
+    """
+
+    input: Any = attr.ib()
+    tms: TileMatrixSet = attr.ib(default=WEB_MERCATOR_TMS)
+    minzoom: int = attr.ib()
+    maxzoom: int = attr.ib()
+
+    reader: Type[BaseReader] = attr.ib(default=Reader)
+    reader_options: Dict = attr.ib(factory=dict)
+
+    ctx: Any = attr.ib(default=rasterio.Env)
+
+    include_asset_types: Set[str] = attr.ib(default=valid_types)
+
+    @minzoom.default
+    def _minzoom(self):
+        return self.tms.minzoom
+
+    @maxzoom.default
+    def _maxzoom(self):
+        return self.tms.maxzoom
+
+    def _get_reader(self, asset_info: AssetInfo) -> Type[BaseReader]:
+        """Get Asset Reader."""
+        asset_type = asset_info.get("type", None)
+
+        if asset_type and asset_type in [
+            "application/x-hdf5",
+            "application/x-hdf",
+            "application/vnd.zarr",
+            "application/x-netcdf",
+            "application/netcdf",
+        ]:
+            return XarrayReader
+
+        return Reader
+
+    def _get_asset_info(self, asset: str) -> AssetInfo:
+        """
+        Validate asset names and return asset's info.
+
+        Args:
+            asset (str): asset name.
+
+        Returns:
+            AssetInfo: Asset info
+
+        """
+        if asset not in self.assets:
+            raise InvalidAssetName(
+                f"{asset} is not valid. Should be one of {self.assets}"
+            )
+
+        asset_info = self.input["assets"][asset]
+
+        url = asset_info["href"]
+        if alternate := stac_config.alternate_url:
+            url = asset_info["alternate"][alternate]["href"]
+
+        info = AssetInfo(url=url, env={})
+
+        if asset_info.get("type"):
+            info["type"] = asset_info["type"]
+
+        if header_size := asset_info.get("file:header_size"):
+            info["env"]["GDAL_INGESTED_BYTES_AT_OPEN"] = header_size  # type: ignore
```

Comment:
it may be worth asking MS PC folks if the file or raster extensions are ever used, because if they're never used this code and the block below are unnecessary.

Reply:
same goes for alternate URL

Reply:
it doesn't hurt to keep this